### ISSUE TITLE
mcplot-pyqtgraph-py H8 output AND templateSANS

### ISSUE THREAD

**Issue by willend** [McStasMcXtrace/McCode#336]
Issue plotting datasets from BNL_H8 - error message along lines of could not load. (backend)

**Comment by climbcat** [comment]
The problem with templateSANS comes from the interpretation of data - there are only three monitors listed in the sweep/base mccode.sim and mccode.dat, but there are in fact four monitor data files in all of the sweep subfolders. This causes an inconsistency to arise in the generated plot graph.

**Comment by climbcat** [comment]
The BNL_H8 data has no mccode.dat file in it, which is right now assumed by the data loader backend Why isn't the mccode.dat file being written during sumulations? Can we expect that it should be?

**Comment by willend** [comment]
It should indeed be written during simulation! Perhaps the scan was done with mcrun-py and we are back in the situation where it does not pick up the Detector: lines from each sim step? Python 3 issue?

**Comment by willend** [comment]
The other thing about 4 vs 3 output data files: If there are 4 files they should all be visible in the mccode.sim of each subfolder?

**Comment by climbcat** [comment]
All four dat files should indeed be visible in the mcstas.sim in each sub folder, but those files are never read by the loader, since we don't have a plot view that shows sub folder simulations. 

What do you mean by the situation where mcrun-py does not pick up on Detector: lines from each sim step??

**Comment by willend** [comment]
OK let me rephrase:
1) mccode.dat should be there, it is the task of the mcrun tool(s) to generate that file (by looking for lines of type Detector: Filename Intensity Err in the stdout from each sim step)
2) I remember a bug that mcrun-py did not generate mccode.dat, but thought this was fixed
3) If this was indeed fixed in the past, it could be that switching to python 3 broke it again

:-)

Peter

**Comment by climbcat** [comment]
Aha ok, understood:)

**Comment by willend** [comment]
Hi again - update!

The issue seems NOT to be with mcrun-py but rather mcrun(.pl). When performing a scan it seems that the perl mcrun only picks up a single Detector: line from monitors that have multiple outputs, resulting in only 3 sets of column data in mcstas.dat even though there were in princple 4 signals.

Example datasets with identical seeds made with each mcrun attached.

I'd better have a look at this one then...

[mcrun_scans.zip](https://github.com/McStasMcXtrace/McCode/files/612887/mcrun_scans.zip)

**Comment by willend** [comment]
Fixed on the perl side

**Comment by climbcat** [comment]
nice! :+1:

**Comment by climbcat** [comment]
BNL H8 uses .psd for its "assumed to be .dat files in the code"! Reopening.

**Comment by climbcat** [comment]
Fixed temporarily by having the loader view .psd files as data files (in addition to .dat files). Opening an issue to generalize this.

**Comment by willend** [comment]
The thing is that nothing can be assumed, the user has full flexibility to choose the file name and suffix... And in the case of monitor ND they can even be generated to be "almost anything" 😊

**Comment by climbcat** [comment]
There are two steps in the loading algorithm, the first establishes what kind of data we are dealing with, the second loads the files listed in mccode.sim (no assumptions). The error occurred in the first step. I am thinking that it should trigger positive on a mccode.sim file plus ANY files beside it, no questions asked.
